lib/internal/repl.js

```diff
@@ -162,62 +160,25 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
         }
       }
     }
-
-    fs.open(historyPath, 'w', onhandle);
-  }
-
-  function onhandle(err, hnd) {
-    if (err) {
-      return ready(err);
-    }
-    repl._historyHandle = hnd;
-    repl.on('line', online);
-
-    // reading the file data out erases it
-    repl.once('flushHistory', function() {
-      repl.resume();
-      ready(null, repl);
-    });
-    flushHistory();
-  }
-
-  // ------ history listeners ------
-  function online() {
-    repl._flushing = true;
-
-    if (timer) {
-      clearTimeout(timer);
-    }
-
-    timer = setTimeout(flushHistory, kDebounceHistoryMS);
+    // Flush history on close
+    repl.on('close', flushHistory);
+    repl.resume();
+    ready(null, repl);
   }
 
   function flushHistory() {
-    timer = null;
-    if (writing) {
-      pending = true;
-      return;
-    }
-    writing = true;
-    const historyData = repl.history.join(os.EOL);
-    fs.write(repl._historyHandle, historyData, 0, 'utf8', onwritten);
-  }
-
-  function onwritten(err, data) {
-    writing = false;
-    if (pending) {
-      pending = false;
-      online();
-    } else {
-      repl._flushing = Boolean(timer);
-      if (!repl._flushing) {
-        repl.emit('flushHistory');
-      }
+    const data = fs.readFileSync(historyPath, 'utf8');
+    const newLines = repl.history.length - initialHistoryLength;
+    let history = repl.history.slice(0, newLines);
+    if (data) {
+      history = history.concat(parseData(data)).slice(0, repl.historySize);
     }
+    const fd = fs.openSync(historyPath, 'a');
+    fs.writeSync(fd, history.join('\n'), 0, 'utf8');
```

Comment:
@addaleax this is strange. I am not seeing this. I'll grant you that a newline terminating the file is good practice, and will address that. But for the record, I am not seeing the behavior you show above. I don't know why..... 

    ~/s/node git:repl-history-file ❯❯❯ git status                                               ✭
    On branch repl-history-file
    nothing to commit, working directory clean
    ~/s/node git:repl-history-file ❯❯❯ git log -n1                                              ✭
    commit f5302b44eeb48039d1e140b988e06b930fdeac5e
    Author: Lance Ball <lball@redhat.com>
    Date:   Tue Jul 26 17:36:59 2016 -0700

        repl: History file improvements

        Address concerns in tests by using common.mustCall.
    ~/s/node git:repl-history-file ❯❯❯ rm ~/.node_repl_history                                  ✭
    remove /Users/lanceball/.node_repl_history? y
    ~/s/node git:repl-history-file ❯❯❯ ./node                                                   ✭
    > 1
    1
    > 2
    2
    > 3
    3
    > .exit
    ~/s/node git:repl-history-file ❯❯❯ ./node                                                   ✭
    > 4
    4
    > 5
    5
    > 6
    6
    > .exit
    ~/s/node git:repl-history-file ❯❯❯ cat ~/.node_repl_history                                 ✭
    .exit
    6
    5
    4
    .exit
    3
    2
    1%